tasks/build/download_node_builds.js

```diff
@@ -1,84 +1,133 @@
+    };
+
+    if (!grunt.file.isDir(platform.nodeDir)) {
+      return false;
+    }
 
-      if (platform.win) {
-        await fromNode(cb => {
-          resp
-          .pipe(createWriteStream(resolve(downloadDir, 'node.exe')))
-          .on('error', cb)
-          .on('finish', cb);
-        });
-      } else {
-        await fromNode(cb => {
-          resp
-          .pipe(createGunzip())
-          .on('error', cb)
-          .pipe(new Extract({ path: downloadDir, strip: 1 }))
-          .on('error', cb)
-          .on('end', cb);
-        });
+    return checkHashFromFile(path, expected).then(([passed, actual]) => {
+      if (passed) {
+        grunt.log.ok(`${platform.name} shasum verified`);
       }
+      else {
+        grunt.log.error(`${platform.name} shasum check failed`);
+      }
+      return passed;
+    });
+  };
 
-      await fromNode(cb => {
-        rename(downloadDir, finalDir, cb);
-      });
-    }());
+  const getNodeBuild = (platform) => {
+    const downloadDir = platform.nodeDir;
+    const file = basename(platform.nodeUrl);
+
+    if (grunt.file.isDir(platform.nodeDir)) {
+      grunt.file.delete(platform.nodeDir);
+    }
+
+    grunt.log.ok(`downloading ${platform.name}`);
+
+    return wreckGet(platform.nodeUrl)
+    .then(([resp, payload]) => {
+      if (resp.statusCode !== 200) {
+        throw new Error(`${platform.nodeUrl} failed with a ${resp.statusCode} response`);
+      }
+      grunt.file.mkdir(downloadDir);
+      return payload;
+    })
+    .then(payload => {
+      return writeFileAsync(resolve(downloadDir, file), payload);
+    });
 
-    activeDownloads.push(platform);
+  };
 
-    var bytes = parseInt(resp.headers['content-length'], 10) || 'unknown number of';
-    var mb = ((bytes / 1024) / 1024).toFixed(2);
-    grunt.log.ok(`downloading ${platform.name} - ${mb} mb`);
+  const extractNodeBuild = async (platform) => {
+    const file = basename(platform.nodeUrl);
+    const path = resolve(platform.nodeDir, file);
+
+    return new Promise((resolve, reject) => {
+      createReadStream(path)
+      .pipe(createGunzip())
+      .on('error', reject)
+      .pipe(new Extract({path: platform.nodeDir, strip: 1}))
+      .on('error', reject)
+      .on('end', resolve);
+    });
   };
 
-  grunt.registerTask('_build:downloadNodeBuilds:start', function () {
-    map(platforms, start).nodeify(this.async());
-  });
+  const start = async (platform) => {
+    let downloadCounter = 0;
+    let isDownloadValid = await checkShaSum(platform);
+
+    if (isDownloadValid) return;
 
-  grunt.registerTask('_build:downloadNodeBuilds:finish', function () {
-    map(activeDownloads, async (platform) => {
-      await platform.downloadPromise;
-      grunt.log.ok(`${platform.name} download complete`);
+    grunt.log.ok('starting download ...');
+    while (!isDownloadValid && (downloadCounter < downloadLimit)) {
```

Comment:
lets say you are on a shitty connection .... your first download went thru ok (linux 32) ... your second download fails (shasum is wrong). if we dont have this limits we would just quit and notify user something went wrong. now user runs build script again ... this time another download failed .... can get quite frustrating.

we retry each failed download automatically (but up to 3 times ... seems a reasonable number) until its succesful.